matching/README.md

```diff
@@ -0,0 +1,991 @@
+Practically, we can define set patterns explicitly in a similar fashion to map
+patterns, except that set patterns do not have values and thus do not generate
+any columns or decision tree nodes corresponding to the value sort of the map.
+The rules are otherwise substantially identical, except for `decomposeSet`,
+which we will now describe:
+
+```
+def decomposeSet(Matrix m, Column j): DT
+  val newO = size : o_j
+  val sk = element sort of s_j
+  if signature(s_j, j) contains Empty:
+    val cases = [(0, compiledCases(m, j)[Empty])]
+    val switch = Switch(newO, Int, cases, compiledDefault(m, j))
+    val f = the function that returns the length of a set
+    return Function(newO, f, [(o_j, s_j)], Int, switch)
+  else if signature(s_j, j) contains Choice:
+    val trueCase = compiledCases(m, j)[Choice]
+    val falseCase = default(m, j)
+    val f1' = the function that removes an element from a set
+    val f2' = the function that returns the next element of an iterator, or None otherwise
+    val f3' = the function that creates an iterator for a set
+    val f1 = Function(rem : o_j, f1', [(o_j, s_j), (key : o_j, sk)], s_j, trueCase)
+    val check = IterHasNext(key : o_j, sk, f1, falseCase)
+    val f2 = Function(key : o_j, f2', [(iter : o_j, Iterator)], sk, check)
+    return Function(iter : o_j, f3', [(o_j, s_j)], Iterator, f3)
+  else:
+    Key(k) is in signature(s_j, j)
+    val trueCase = compiledCases(m, j)[Key(k)]
+    val falseCase = compiledDefault(m, j)
+    val check = Switch(value : pat(k) : o_j, Bool,
+        [(true, trueCase)], falseCase)
+    val f = the function that removes an element from a set
+    val g = the function that tests set membership
+    val f1 = Function(rem : pat(k) : o_j, f, [(o_j, s_j), (newO, sk)], s_j, check)
+    val f2 = Function(value : pat(k) : o_j, g, [(o_j, s_j), (newO, sk)], Bool, f1)
+    return Pattern(newO, sk, k, f2)
+```
+
+### Heuristics
+
+We have now fully described how to generate decision trees that successfully
+match patterns of the `Pattern` sort described above. However, as is well
+known, not all decision trees for a given problem are equal. In particular, the
+algorithm makes no particular assumptions about which column is decomposed for
+any given matrix, nor about which map key is chosen to decompose for any given
+map column signature. Poor choices here can be catastrophic. As a result,
+we need to redefine the q, b, a, and L heuristics used by K in light of the
+differences between the Maranget paper and our own design, as well as defining
+the `bestKey` function.
+
+First, we start from the assumption we want to avoid set and map choices
+whenever possible, as these are very expensive. Thus, we define the concept of
+a *valid* or *invalid* column. A column over list or regular sorts is always
+valid. A column over map or set sorts is valid if all of its map patterns are
+of the form Map [], or if there exists a key k in a map pattern in the column
+such that k is bound.
```

Comment:
Valid here captures the idea of columns that don't require us to perform an expensive choice operation; there might be a better way to name it but I can't think of one off the top of my head.